gpflow/experimental/check_shapes/checker.py

```diff
@@ -185,6 +185,37 @@ def finished(self) -> bool:
         return self.expected_begin >= self.expected_end
 
 
+@dataclass
+class _VariableState:
+    """
+    Structure of stuff we need to know about each variable.
+    """
+
+    uses: List[Tuple[ParsedTensorSpec, ErrorContext]]
+    """ List of specs where this variable is used. """
+
+    observed_dim: Optional[_ObservedDim]
+    """
+    Observed size of this variable, if the variable is rank-1.
+
+    Set this `None` if this variable is varrank.
+    """
+
+    observed_dims: Optional[_ObservedDims]
+    """
+    Observed shape of this variable, if the variable is varrank.
+
+    Set this `None` if this variable is rank-1.
+    """
+
+    waiting_for_varrank: Set[_ShapeCheck]
+    """ Checks that are waiting for the rank of this variable to be determined. """
+
+    @staticmethod
+    def new() -> "_VariableState":
+        return _VariableState([], None, None, set())
```

Comment:
Why not use field default values instead (using `field(default_factory=list)` etc for the mutable fields)? And then you can just write `DefaultDict(_VariableState)`.

Reply:
> Why not use field default values instead (using `field(default_factory=list)` etc for the mutable fields)? And then you can just write `DefaultDict(_VariableState)`.

That probably is more elegant. I'll do that.